lib/pbench/server/api/resources/query_apis/dataset.py

```diff
@@ -0,0 +1,186 @@
+                context["owner"] = es_doc["owner"] = owner
+
+            # Generate a series of bulk update documents, which will be passed to
+            # the Elasticsearch bulk helper.
+            #
+            # Note that the "doc" specifies explicit instructions for updating only
+            # the "access" field of the "authorization" subdocument: no other data
+            # will be modified.
+            for index, ids in map.items():
+                for id in ids:
+                    yield {
+                        "_op_type": action,
+                        "_index": index,
+                        "_id": id,
+                        "doc": {"authorization": es_doc},
+                    }
+
+        elif action == "delete":
+            current_app.logger.info("Starting delete operation for dataset {}", dataset)
+            sync = Sync(logger=current_app.logger, component=OperationName.DELETE)
+            sync.update(dataset=dataset, state=OperationState.WORKING)
+
+            # Generate a series of bulk delete documents, which will be passed to
+            # the Elasticsearch bulk helper.
+
+            for index, ids in map.items():
+                for id in ids:
+                    yield {"_op_type": action, "_index": index, "_id": id}
+
+    def complete(
+        self, dataset: Dataset, context: ApiContext, summary: JSONOBJECT
+    ) -> None:
+        """
+        Complete the publish operation by updating the access of the Dataset
+        object.
+
+        Note that an exception will be caught outside this class; the Dataset
+        object will remain in the previous state to allow a retry.
+
+        Args:
+            dataset: Dataset object
+            context: CONTEXT dictionary
+            summary: summary of the bulk operation
+                ok: count of successful updates
+                failure: count of failures
+        """
+        auditing: dict[str, Any] = context["auditing"]
+        attributes = auditing["attributes"]
+        state = OperationState.FAILED
+        message = "Unable to update some indexed documents"
+        if context["attributes"]["action"] == "update":
+            if summary["failure"] == 0:
+                state = OperationState.OK
+                message = None
+                access = context.get("access")
+                if access:
+                    attributes["access"] = access
+                    dataset.access = access
+                owner = context.get("owner")
+                if owner:
+                    attributes["owner"] = owner
+                    dataset.owner_id = owner
+                dataset.update()
+            context["sync"].update(dataset=dataset, state=state, message=message)
+        elif context["attributes"]["action"] == "delete":
+            if summary["failure"] == 0:
+                cache_m = CacheManager(self.config, current_app.logger)
+                cache_m.delete(dataset.resource_id)
+                dataset.delete()
```

Comment:
Looking at the restructured code, I realize there's something I missed. A failed `DELETE` will leave the delete operation in `WORKING` state. We don't need to update anything on *success* because the dataset is gone; but on failure we should really do a `sync.error(message)` ... even though the common `message` refers to `update`, that's probably fine... although that could easily be fixed by using `f"Unable to {action} some indexed documents"` up at line 167.